File: Psiphon-Browser/LogViewController.m
/*
 * Copyright (c) 2016, Psiphon Inc.
 * All rights reserved.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 *
 */

#import <Foundation/Foundation.h>
#import "LogViewController.h"
#import "FeedbackViewController.h"


@interface LogViewController ()

@property (strong, nonatomic) NSMutableArray *logQueue;
@property (strong, nonatomic) UIFont *messageFont;

@end


@implementation LogViewController
{
    __weak IBOutlet UISegmentedControl *segmentedControl;
    IBOutlet UITableView *loggingTableView;
    enum {
        OPEN_BROWSER_BUTTON = 0,
        FEEDBACK_BUTTON
    };
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Disable 'open browser' button until tunnels are established
    [segmentedControl setEnabled:NO forSegmentAtIndex:0];
    
    [[NSNotificationCenter defaultCenter] addObserver:self
                                             selector:@selector(receivedNewLogEntryNotification:)
                                                 name:@"NewLogEntryPosted"
                                               object:nil];
    
    _messageFont = [UIFont systemFontOfSize:14];
}

-(void) receivedNewLogEntryNotification:(NSNotification*) aNotification
{
    if ([[aNotification name] isEqualToString: @"NewLogEntryPosted"])
    {
        NSDictionary *userInfo = aNotification.userInfo;
        if( userInfo )
        {
            NSString* logEntry = [userInfo objectForKey:@"LogEntryKey"];
            if(logEntry)
            {
                if (_logQueue == nil) {
                    _logQueue = [[NSMutableArray alloc] init];
                }
                [_logQueue addObject: logEntry];
                
                [loggingTableView setContentOffset:CGPointMake(0, CGFLOAT_MAX)];
                
                [loggingTableView reloadData];
            }
            
            // Observe 'Connected' message
            if ([logEntry  isEqual: @"Connected"]) {
                [segmentedControl setEnabled:YES forSegmentAtIndex:0];
            }
        }
    }
}

- (IBAction)segmentedControlButtonPressed:(id)sender {
    UISegmentedControl *cntrl = (UISegmentedControl *)sender;
    
    if (cntrl.selectedSegmentIndex == OPEN_BROWSER_BUTTON) {
        PsiphonBrowserViewController *browserViewController = [[PsiphonBrowserViewController alloc] initWithNibName:nil bundle:nil];
        UINavigationController *navController = [[UINavigationController alloc] initWithRootViewController:browserViewController];
        navController.navigationBar.hidden = YES;
        
        AppDelegate *appDelegate = (AppDelegate*)[[UIApplication sharedApplication]delegate];
        
        NSString *homepage = [appDelegate getHomepage];
        
        NSString *pageToLoad = homepage ? homepage : @"";
        
        [self presentViewController:navController animated:YES completion:^{
            [browserViewController addTabWithAddress:pageToLoad];
        }];
    } else if (cntrl.selectedSegmentIndex == FEEDBACK_BUTTON) {
        UIStoryboard *sb = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
        UIViewController *vc = [sb instantiateViewControllerWithIdentifier:@"FeedbackViewController"];
        vc.modalTransitionStyle = UIModalTransitionStyleFlipHorizontal;
        [self presentViewController:vc animated:YES completion:nil];
    } else {
        // Do nothing
    }
    [cntrl setSelectedSegmentIndex:-1];
}

- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section
{
    return [_logQueue count];
}

- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath
{
    static NSString *CellIdentifier = @"Cell";
    
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:CellIdentifier];
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:CellIdentifier];
        cell.textLabel.lineBreakMode = NSLineBreakByWordWrapping;
        cell.textLabel.numberOfLines = 0;
        cell.textLabel.font = _messageFont;
    }
    
    // Configure the cell...
    NSString *message = [_logQueue objectAtIndex:indexPath.row];
    
    cell.textLabel.text = message;
    cell.textLabel.font = _messageFont;
    cell.textLabel.numberOfLines = 0;
    
    return cell;
}

// Prevent logs from getting truncated in tableView UI
- (CGFloat)tableView:(UITableView *)tableView heightForRowAtIndexPath:(NSIndexPath *)indexPath
{
    NSString *cellText = [_logQueue objectAtIndex:indexPath.row];
    NSAttributedString *attributedText =
    [[NSAttributedString alloc]
     initWithString:cellText
     attributes:@
     {
     NSFontAttributeName: _messageFont
     }];
    CGRect rect = [attributedText boundingRectWithSize:CGSizeMake(tableView.bounds.size.width, CGFLOAT_MAX)
                                               options:NSStringDrawingUsesLineFragmentOrigin
                                               context:nil];
    return rect.size.height + 20;
}

@end

File: Psiphon-Browser/LogViewController.swift
//
//  File.swift
//  Psiphon-Browser
//
//  Created by Miro Kuratczyk on 2016-08-25.
//  Copyright © 2016 Psiphon. All rights reserved.
//

import Foundation
import UIKit

class LogViewController: UIViewController, UITableViewDelegate, UITableViewDataSource {
    @IBOutlet weak var tableView: UITableView!
    @IBOutlet weak var segmentedControl: UISegmentedControl!
    
    var displayedLogs: [String] = []
    
    enum Buttons: Int {
        case OPEN_BROWSER = 0
        case FEEDBACK
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()

        self.tableView.register(UITableViewCell.self, forCellReuseIdentifier: "cell")
        
        segmentedControl.setEnabled(false, forSegmentAt: 0)
        
        // Add observers
        NotificationCenter.default.addObserver(self, selector: #selector(receivedConnectionEstablished), name: NSNotification.Name.init(rawValue: Constants.Notifications.ConnectionEstablished), object: nil)
        NotificationCenter.default.addObserver(self, selector: #selector(receivedUpdateLogsNotification), name: NSNotification.Name.init(rawValue: Constants.Notifications.DisplayLogEntry), object: nil)
    }
    
    func receivedConnectionEstablished(aNotification: Notification) {
        let result = aNotification.withName(name: Constants.Notifications.ConnectionEstablished)
        
        switch result {
        case .Value(_):
            self.segmentedControl.setEnabled(true, forSegmentAt: 0)
        case let .Error(error):
            print("Error receiving notification: " + String(error))
        }
    }
    
    func receivedUpdateLogsNotification(aNotification: Notification) {
        let result = aNotification.withName(name: Constants.Notifications.DisplayLogEntry)
        
        switch result {
        case .Value(_):
            displayedLogs = PsiphonData.sharedInstance.getDiagnosticLogs()
            tableView.reloadData()
            scrollToBottom()
        case let .Error(error):
            print("Error receiving notification: " + String(error))
        }
    }
    
    func scrollToBottom() {
        let bottomOffset: CGPoint = CGPoint(x: 0, y: tableView.contentSize.height - tableView.frame.size.height)
        tableView.setContentOffset(bottomOffset, animated: true)
    }
    
    @IBAction func segmentedControlButtonPressed(_ sender: UISegmentedControl) {
        if (sender.selectedSegmentIndex == Buttons.OPEN_BROWSER.rawValue) {
            
            let appDelegate = UIApplication.shared.delegate as! AppDelegate
            let homepage = appDelegate.getHomepage()
            let pageToLoad = homepage != nil ? homepage! : ""
            
            let browserVC = PsiphonBrowserViewController()
            let navCntrl = UINavigationController.init(rootViewController: browserVC)
            navCntrl.navigationBar.isHidden = true
            self.present(navCntrl, animated: true, completion: { browserVC.addTab(withAddress: pageToLoad)} )
            
        } else if (sender.selectedSegmentIndex == Buttons.FEEDBACK.rawValue) {
            let vc = self.storyboard?.instantiateViewController(withIdentifier: "FeedbackViewController")
            self.present(vc!, animated: true, completion: nil)
        } else {
            // Unreachable: Do nothing
        }
        
        scrollToBottom()
        sender.selectedSegmentIndex = -1
    }
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.displayedLogs.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell:UITableViewCell = self.tableView.dequeueReusableCell(withIdentifier: "cell")! as UITableViewCell
        
        cell.textLabel?.text = self.displayedLogs[indexPath.row]
        cell.textLabel?.lineBreakMode = NSLineBreakMode.byWordWrapping
        cell.textLabel?.font = UIFont.systemFont(ofSize: 14)
        cell.textLabel?.numberOfLines = 0
        
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
       print("You selected cell #\(indexPath.row)!") // TODO: remove
    }
}
